plugins/out_stackdriver/stackdriver_http_request.c

```diff
@@ -0,0 +1,453 @@
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2020 The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#include <fluent-bit/flb_regex.h>
+#include "stackdriver.h"
+#include "stackdriver_http_request.h"
+
+typedef enum {
+    NO_HTTPREQUEST = 1,
+    HTTPREQUEST_EXISTS = 2
+} http_request_status;
+
+void init_http_request(struct http_request_field *http_request)
+{
+    http_request->latency = flb_sds_create("");
+    http_request->protocol = flb_sds_create("");
+    http_request->referer = flb_sds_create("");
+    http_request->remoteIp = flb_sds_create("");
+    http_request->requestMethod = flb_sds_create("");
+    http_request->requestUrl = flb_sds_create("");
+    http_request->serverIp = flb_sds_create("");
+    http_request->userAgent = flb_sds_create("");
+
+    http_request->cacheFillBytes = 0;
+    http_request->requestSize = 0;
+    http_request->responseSize = 0;
+    http_request->status = 0;
+
+    http_request->cacheHit = FLB_FALSE;
+    http_request->cacheLookup = FLB_FALSE;
+    http_request->cacheValidatedWithOriginServer = FLB_FALSE;
+}
+
+void destroy_http_request(struct http_request_field *http_request)
+{
+    flb_sds_destroy(http_request->latency);
+    flb_sds_destroy(http_request->protocol);
+    flb_sds_destroy(http_request->referer);
+    flb_sds_destroy(http_request->remoteIp);
+    flb_sds_destroy(http_request->requestMethod);
+    flb_sds_destroy(http_request->requestUrl);
+    flb_sds_destroy(http_request->serverIp);
+    flb_sds_destroy(http_request->userAgent);
+}
+
+void add_http_request_field(struct http_request_field *http_request, 
+                            msgpack_packer *mp_pck)
+{    
+    msgpack_pack_str(mp_pck, 11);
+    msgpack_pack_str_body(mp_pck, "httpRequest", 11);
+
+    if (flb_sds_is_empty(http_request->latency) == FLB_TRUE) {
+        msgpack_pack_map(mp_pck, 14);
+    }
+    else {
+        msgpack_pack_map(mp_pck, 15);
+
+        msgpack_pack_str(mp_pck, 7);
+        msgpack_pack_str_body(mp_pck, "latency", 7);
+        msgpack_pack_str(mp_pck, flb_sds_len(http_request->latency));
+        msgpack_pack_str_body(mp_pck, http_request->latency, 
+                              flb_sds_len(http_request->latency));
+    }
+
+    /* String sub-fields */
+    msgpack_pack_str(mp_pck, 13);
```

Comment:
can you pack sub-fields in alphabetical order too? Packing them by value type is fine for me.

Reply:
Done